File: packages/ts/src/v1/validate.ts
import type { Issue, Result, OcfDoc } from "../types.js";
import { assemble } from "../types.js";
import { schemaLevel } from "../schema-level.js";
import { buildContext } from "./context.js";
import { possessionByFrame } from "./possession.js";
import { referenceRules } from "./rules/references.js";
import { possessionRules } from "./rules/possession-rules.js";
import { coherenceRules } from "./rules/coherence.js";
import { qualityRules } from "./rules/quality.js";
import { schemaCheck } from "../schema-version.js";
import { makeIssue } from "../codes.js";

export function validate(doc: OcfDoc): Result {
  if (typeof doc !== "object" || doc === null || Array.isArray(doc)) {
    throw new TypeError("validate: expected an object (parsed OCF document)");
  }
  const check = schemaCheck(doc);

  const issues: Issue[] = [];
  // Document needs a newer minor than we bundle: warn, then best-effort validate.
  if (check.outdated) {
    issues.push(makeIssue("VALIDATOR_MAYBE_OUTDATED", "/meta/min_schema_version", {
      required: check.block.requiredByDoc, bundled: check.block.validatedAgainst,
    }));
  }

  const level0 = schemaLevel(doc);
  if (level0.length > 0) return assemble([...issues, ...level0], check.block); // stop on schema/legacy failure

  const ctx = buildContext(doc);
  const states = possessionByFrame(doc);
  issues.push(
    ...referenceRules(doc, ctx),
    ...possessionRules(doc, ctx, states),
    ...coherenceRules(doc, ctx),
    ...qualityRules(doc, ctx),
  );
  return assemble(issues, check.block);
}

File: packages/py/ocf_validator/v1/validate.py
from ..codes import make_issue
from ..schema_level import schema_level
from ..schema_version import schema_check
from ..types import Issue, Result, assemble
from .context import build_context
from .possession import possession_by_frame
from .rules import (
    coherence_rules,
    possession_rules,
    quality_rules,
    reference_rules,
)


def validate(doc) -> Result:
    if not isinstance(doc, dict):
        raise TypeError("validate: expected a dict (parsed OCF document)")
    check = schema_check(doc)

    issues: list[Issue] = []
    # Document needs a newer minor than we bundle: warn, then best-effort validate.
    if check["outdated"]:
        issues.append(make_issue("VALIDATOR_MAYBE_OUTDATED", "/meta/min_schema_version", {
            "required": check["block"]["requiredByDoc"], "bundled": check["block"]["validatedAgainst"],
        }))

    level0 = schema_level(doc)
    if level0:
        return assemble(issues + level0, check["block"])
    ctx = build_context(doc)
    states = possession_by_frame(doc)
    issues.extend(reference_rules(doc, ctx))
    issues.extend(possession_rules(doc, ctx, states))
    issues.extend(coherence_rules(doc, ctx))
    issues.extend(quality_rules(doc, ctx))
    return assemble(issues, check["block"])
